### DISCUSSION TITLE
CopyToOutputDirectory does not work for AvaloniaXaml resources

### DISCUSSION THREAD

**Discussion by tisis2** [AvaloniaUI/Avalonia#17340] [Q&A]
**Describe the bug**
When you add the CopyToOutputDirectory Attribute to any of your AvaloniaXaml resources it wont be respected and not be placed in the output directory.
personly i had this issue when i implemented a custom analyzer for axaml files and my translation markup extension. i wanted to write unit tests for this analyzer and place some axaml files into the test project that would be made available for the test by copy them to the build dir...

**To Reproduce**

1. create a avalonia project
2. set CopyToOutputDirectory on any axaml file

**Expected behavior**
the xaml is copied to output dir like when this property is set on other resources

**Additional context**
the workaround found
```
<AvaloniaXaml Remove="*.axaml" />
<Content Include="*.axaml">
    <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
</Content>
```

but now i often get the exception during build
```
MSB4018
The "GenerateAvaloniaResourcesTask" task failed unexpectedly.
System.IO.FileNotFoundException: Could not find file '....Analyzer.Tests\bin\Debug\net6.0\TestControls\....axaml'.
File name: '....Analyzer.Tests\bin\Debug\net6.0\TestControls\....axaml'
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileInfo.get_Length()
   at Avalonia.Build.Tasks.GenerateAvaloniaResourcesTask.Source..ctor(String relativePath, String root)
   at Avalonia.Build.Tasks.GenerateAvaloniaResourcesTask.<BuildResourceSources>b__22_0(ITaskItem r)
   at System.Linq.Enumerable.WhereSelectArrayIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Avalonia.Build.Tasks.GenerateAvaloniaResourcesTask.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()	

...\.nuget\packages\avalonia\0.10.18\buildTransitive\AvaloniaBuildTasks.targets	69
```

most of the time this disappears on a second build... but still annoying that two builds are needed to get a working build

**Comment by samtun** [comment]
I just ran into the same issue.
The proposed workaround does not work for me. The application crashes directly with this exception (which sounds kind of expected...)

```
Avalonia.Markup.Xaml.XamlLoadException:
'No precompiled XAML found for MyApp.App, make sure to specify x:Class and include your XAML file as AvaloniaResource'
```

___

The only other workaround I found is copying the files manually in a PostBuild event like so:

```
<Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="copy $(ProjectDir)Some\Directory\MyControl.axaml $(OutDir)Some\Directory\" />
</Target>
```

This works fine for my use case of reading the axaml as string in the program itself.

A fix would still be much appreciated.

**Comment by maxkatz6** [comment]
Can you clarify what are you trying to do?
AvaloniaXaml is an embedded resource, it can't be copied to the output dir by definition.
If you want it to be copied (as a duplicate), you likely need to do something like this:
```
<AvaloniaXaml Include="MyFile" />
<Content Include="MyFile"  CopyToOutputDirectory="PreserveNewest" />
```

I.e. `Content` item group is specifically responsible for non-embedded files in the output directory.
If you remove item from `AvaloniaXaml`, it's only expected to get `XamlLoadException` later, as normally Avalonia needs XAML files to be embedded (as XAML before compilation, and as IL code after compilation).

**Reply by samtun** [reply]
I'm not that deep into what file is what to dotnet and did not expect to require an explicit Include tag here.
That actually works! Thank you so much.
